src/util/heap/freelistpageresource.rs

```diff
@@ -181,27 +144,21 @@ impl<VM: VMBinding> PageResource<VM> for FreeListPageResource<VM> {
 impl<VM: VMBinding> FreeListPageResource<VM> {
     pub fn new_contiguous(start: Address, bytes: usize, vm_map: &'static dyn VMMap) -> Self {
         let pages = conversions::bytes_to_pages_up(bytes);
-        let common_flpr = {
-            let common_flpr = Box::new(CommonFreeListPageResource {
-                free_list: vm_map.create_parent_freelist(start, pages, PAGES_IN_REGION as _),
-                start,
-            });
-            // `CommonFreeListPageResource` lives as a member in space instances.
-            // Since `Space` instances are always stored as global variables, so it is okay here
-            // to turn `&CommonFreeListPageResource` into `&'static CommonFreeListPageResource`
-            unsafe {
-                vm_map.bind_freelist(common_flpr.as_ref() as *const _);
-            }
-            common_flpr
+        let free_list = vm_map.create_parent_freelist(start, pages, PAGES_IN_REGION as _);
+        let actual_start = if let Some(rmfl) = free_list.downcast_ref::<RawMemoryFreeList>() {
```

Comment:
We could let `dyn FreeList` to return a value based on the implementation rather than downcasting the type here. The `FreeList` implementation is internal to the `VMMap`, and we should not break the abstraction and let `PageResource` do things based on the `FreeList` implementation.

Reply:
I don't think `FreeList` is an implementation detail of `VMMap`.  It is a very general free list abstraction, and has two implementations, namely `IntArrayFreeList` and `RawMemoryFreeList`.  In this sense, I don't think `trait FreeList` should contain a method that returns the number of bytes it occupies in the space, because a `FreeList` is not necessarily used by a space.  It can be used to allocate any kind of resources.

But it is true that `VMMap` *selects* the implementation.  It is `VMMap::create_freelist` that decide whether to reserve a proportion of memory at the beginning of the space for the `RawMemoryFreeList`.  I think it is natural for `VMMap::create_freelist` to return the number of bytes reserved for the `FreeList` implementation alongside the `dyn FreeList`, i.e. it should return a tuple.

In my previous version of the PR, I introduced the [CreateFreeListResult](https://github.com/mmtk/mmtk-core/pull/925/files#diff-33c211ce4d7709987021ccf3ac97406730facd09ab61c5fae0e8c8a15a284489R7) type:

```rust
pub struct CreateFreeListResult {
    pub free_list: Box<dyn FreeList>,
    pub space_displacement: usize,
}
```

I think I can move it to this PR, too.  If that's too verbose, I can replace it with a simple tuple instead.

Reply:
Returning an offset from the create functions sounds good.

Reply:
I made `create_freelist` return `CreateFreeListResult` so that `FreeListPageResource` no longer performs downcast.